Feature/10up toolkit 3.0.0

nicholasio opened the pull request:
<!--
### Requirements

Filling out the template is required.  Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change

<!--
We must be able to understand the design of your change from this description.  If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion.  Also including any benefits that will be realized by the code change will be helpful.  Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.  Please include screenshots (if appropriate).
-->

<!-- Enter any applicable Issues here. Example: -->
Closes #

This updates `wp-scaffold` to work with 10up-toolkit@3.0.0 and HMR & Fast Refresh.

It basically removes all css entry points from `package.json` and imports them via a corresponding js file.

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected. -->

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

<!--
What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., commands you ran, text you typed, buttons you clicked) and describe the results you observed.
-->

- Clone this branch.
- Make sure to remove node_modules and start from a clean state.
- Run `npm install` from the root
- Set SCRIPT_DEBUG to true
- `npm run watch`
- Edit any CSS file (admin-style.css, frontend.css, editor-styles-overrides.css, styleguide.css etc...) Changes should hot reload
- Edit the example block css and react code. Changes should hot reload.
- 
### Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests passed.

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Changelog Entry

<!--
Add sample CHANGELOG.md entry for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.
Example:
> Added - New feature
> Changed - Existing functionality
> Deprecated - Soon-to-be removed feature
> Removed - Feature
> Fixed - Bug fix
> Security - Vulnerability
-->

### Credits

<!-- Please list any and all contributors on this PR and any linked issue so that they can be added to this projects CREDITS.md file. -->
Props @

nicholasio committed:
Updating 10up-toolkit

nicholasio committed:
WIP: HMR set up

nicholasio committed:
Reworking frontend css

nicholasio committed:
Merge branch 'trunk' into feature/10up-toolkit-2.2.0

nicholasio committed:
Initial pass at integration HMR & fast refresh into wp-scaffold

nicholasio committed:
Fixing styleguide path

nicholasio committed:
Updating loading of dist/fast-refresh.php

nicholasio committed:
Removing unnecessary code

nicholasio committed:
Refactoring plugin to the new structure

nicholasio committed:
Final touches

nicholasio committed:
Removing unnecessary files

nicholasio committed:
Merge branch 'feature/npm-workspaces' into feature/10up-toolkit-2.2.0

nicholasio committed:
Fix Utility file in the plugin

nicholasio committed:
fixing constants in plugin utility

fabiankaegy commented on mu-plugins/10up-plugin/package.json:
```diff
@@ -15,13 +15,12 @@
   },
   "devDependencies": {
-    "10up-toolkit": "^2.2.0-alpha.6"
+    "10up-toolkit": "^2.2.0-alpha.7"
```
I think this should only get merged once the actual 2.2.0 version has been released

nicholasio replied:
Yep, I just wanted to get this reviewed ahead of final release.

fabiankaegy commented on themes/10up-theme/package.json:
```diff
@@ -15,13 +15,12 @@
   },
   "devDependencies": {
-    "10up-toolkit": "^2.2.0-alpha.6"
+    "10up-toolkit": "^2.2.0-alpha.7"
```
Same here

fabiankaegy commented on mu-plugins/10up-plugin/package.json:
```diff
@@ -15,20 +15,16 @@
   },
   "devDependencies": {
-    "10up-toolkit": "^2.1.0"
+    "10up-toolkit": "^2.2.0-alpha.8"
```
Same here

fabiankaegy commented on mu-plugins/10up-plugin/package.json:
```diff
@@ -15,20 +15,16 @@
       "frontend": "./assets/js/frontend/frontend.js",
-      "shared": "./assets/js/shared/shared.js",
-      "admin-style": "./assets/css/admin/admin-style.css",
-      "shared-style": "./assets/css/shared/shared-style.css",
```
I would love to better understand how the entry points now get generated. Is is depending on the name of the JS file they were imported into? Or is it depending on the name of the stylesheet that gets imported? 

There will be cases where a block for example has two separate stylesheets. One for the editor and one for the fronend. They should not be merged into the same stylesheet. 

I think `@wordpress/scripts` handles this by allowing for `style.css` or `editor.css`. And depending on which it is they get put into two separate files.

Additionally, it would be great for the next iteration of all this to allow for per block stylesheet generation. That way there could be a `style.css` and `editor.css` for every single block.

nicholasio replied:
The entrypoint gets generated based on the js file importing them.

So anything imported by `frontend.js` will get bundled in a `frontend.css` including block css etc.

I'm happy to expand this a bit further.

fabiankaegy replied:
Okay yeah, I think this is fine for now and makes sense. In the future when we want to tackle the individual stylesheets per block we will need to differentiate between editor and frontend styles. We can use `@wordpress/scripts` as an inspiration there.

nicholasio replied:
Sounds good.

We could expand this as part of the work to automatically detect block.json files in 10up-toolkit

fabiankaegy commented on themes/10up-theme/includes/blocks.php:
```diff
@@ -113,6 +113,16 @@ function blocks_editor_styles() {
 		Utility\get_asset_info( 'editor-style-overrides', 'version' )
 	);
 
+	if ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) {
```
I could also see using `wp_get_environment_type( 'local' )` here.

nicholasio replied:
That could work too, but HMR only works when SCRIPT_DEBUG is set to true and this enqueue only exists for the purpose of HMR ... so even if the env is set to local the SCRIPT_DEBUG constant could still be false.

fabiankaegy commented on themes/10up-theme/includes/blocks/example-block/edit.js:
```diff
@@ -4,16 +4,19 @@
 import { RichText, useBlockProps } from '@wordpress/block-editor';
 
+// Importing the block's css via JS will enable hot reloading for css
+import './index.css';
```
As part of this, we should rename the `index.css` file to `editor.css`. It currently is confusing and especially now with this new comment it makes it sound as if all the block styles should be written in that file.

nicholasio replied:
Sure, I'll rename this to editor.css

nicholasio replied:
Done!

fabiankaegy commented on themes/10up-theme/includes/core.php:
```diff
@@ -104,6 +108,10 @@ function scripts() {
+	 * Enqueuing shared.js is required to get css hot reloading working in the frontend
+	 * If you're not shipping any shared js wrap this enqueue in a SCRIPT_DEBUG check.
+	 */
 	/*
```
@nicholasio I think you forgot to uncomment this code here.

nicholasio replied:
So the shared style were already commented out in the scaffold so I left commented out.

fabiankaegy commented on themes/10up-theme/package.json:
```diff
@@ -15,24 +15,19 @@
   },
   "devDependencies": {
-    "10up-toolkit": "^2.1.0"
+    "10up-toolkit": "^2.2.0-alpha.8"
```
Again I think we should only merge with the final version.

nicholasio committed:
renaming blocks editor style to editor.css

nicholasio committed:
Updating toolkit to the released version

nicholasio committed:
Changing default ports

joesnellpdx commented:
@nicholasio 

I tried to run a quick test - and maybe there is some user error, but here are my findings:

To test the toolkit 3.0.0 MR... how much setup do I need to get it to work?

- I removed all `node_modules` folders
-I have node 16.14.0 and npm 7.24.2
- I added the them devurl: "devURL": "https://10up-scaffold.test", (based on docs, using .test I may not need this .. so tested with and without)
- I'm running everything from the root
- I'm testing here `https://10up-scaffold.test/`  (I don't think I need a unique URL)

If I run `npm run watch` ... 
I change something in css... I see it compile... but have to refresh the page to see the change.
JS - files are not being enqueued (although I see `frontend.js` and `frontend.asset.php` in `/dist`)  so I can't test JS
Same is true if I run `npm run watch:theme`

^^ tried in the plugin as well - same results

FYI: there is still a `lerna clean` command in root `package.json` 

I don't have the bandwidth to debug currently, but wanted to give you an update. 

== 

I didn't test anything else.

nicholasio commented:
> @nicholasio
> 
> I tried to run a quick test - and maybe there is some user error, but here are my findings:
> 
> To test the toolkit 3.0.0 MR... how much setup do I need to get it to work?
> 
> * I removed all `node_modules` folders
> * I have node 16.14.0 and npm 7.24.2
> * I added the them devurl: "devURL": "https://10up-scaffold.test", (based on docs, using .test I may not need this .. so tested with and without)
> * I'm running everything from the root
> * I'm testing here `https://10up-scaffold.test/`  (I don't think I need a unique URL)
> 
> If I run `npm run watch` ... I change something in css... I see it compile... but have to refresh the page to see the change. JS - files are not being enqueued (although I see `frontend.js` and `frontend.asset.php` in `/dist`) so I can't test JS Same is true if I run `npm run watch:theme`
> 
> ^^ tried in the plugin as well - same results
> 
> FYI: there is still a `lerna clean` command in root `package.json`
> 
> I don't have the bandwidth to debug currently, but wanted to give you an update.
> 
> ==
> 
> I didn't test anything else.

Did you set SCRIPT_DEBUG to true?
`define('SCRIPT_DEBUG', true);`

joesnellpdx commented:
> Did you set SCRIPT_DEBUG to true? `define('SCRIPT_DEBUG', true);`

OMG! Dumb. That did it. Let take it for a deeper spin later today. Thanks!

joesnellpdx reviewed (approved):
@nicholasio 

Approved by me!

Only change maybe is removing `"clean": "lerna clean",` from the root `package.json`

nicholasio committed:
Removing lerna clean command and updating README.md

nicholasio committed:
Merge branch 'trunk' into feature/10up-toolkit-2.2.0

nicholasio commented:
> @nicholasio
> 
> Approved by me!
> 
> Only change maybe is removing `"clean": "lerna clean",` from the root `package.json`

Done!

joesnellpdx commented:
@devinle @fabiankaegy 

Are either of. you able to take a pass by EOW? If not, I recommend @nicholasio go ahead and merge.

fabiankaegy reviewed (approved):
@joesnellpdx I've already given my approval when we tested it in a slack conversation  👍 this is approved from my end  👍

nicholasio commented:
Thanks everyone for the help reviewing and testing this!
